File: examples/remote_control/control_client.ts
/**
 * Example external control application for the siphon control plane — TypeScript SDK.
 *
 * Drives live B2BUA calls a script hands over with `call.handover("ivr-app")` (the
 * ARI *Stasis* model) over siphon's control WebSocket, using the `@siphon-project/control`
 * SDK. The SDK owns the wire completely: no `ws`, no manual JSON, no request-id
 * bookkeeping — you get a `Call` whose verbs read like an in-process siphon script
 * (`call.answer()` / `call.transfer(...)` / `call.hangup()`).
 *
 * Two connection modes, same wire (subprotocol `siphon-control.v1`):
 *
 *   - **outbound per-call-connect (the default)** — this app is a WebSocket
 *     *server* (`SipServer`); siphon dials it once per handed-over call and the
 *     accepting socket owns that call. No `hello` — the first frame is
 *     `StasisStart`. Use this for multi-pod / autoscaled controllers (siphon
 *     always dials *out*, so the "which pod owns the call" affinity problem
 *     cannot arise).
 *   - **inbound persistent** — this app is a WebSocket *client* (`SipClient`)
 *     that connects in to `control.listen` and owns calls assigned to it. It
 *     sends a `hello` and `resync`s to re-attach its calls after a reconnect.
 *
 * `SipClient` / `SipServer` are the SIP facade over the generic `ControlClient` /
 * `ControlServer` core; both expose the same `onCall(handler)` + `Call` verbs.
 *
 * Select the mode with `SIPHON_CONTROL_MODE` (`outbound` | `inbound`).
 *
 * Install the SDK:
 *   npm i @siphon-project/control            # once published
 *   # this example already depends on it via a file: path to the sibling package,
 *   # so `npm install` here wires it up (build the SDK first — see README.md).
 *
 * Usage:
 *   npm install
 *   # outbound (default): siphon dials this server
 *   SIPHON_CONTROL_BIND=0.0.0.0:8443 IVR_APP_TOKEN=changeme-dev-token npm start
 *   # inbound: this app dials siphon
 *   SIPHON_CONTROL_MODE=inbound IVR_APP_TOKEN=changeme-dev-token npm start
 *
 * See README.md for the matching siphon `control:` config and handover script.
 */
import { SipClient, SipServer, ControlError } from "@siphon-project/control";
import type { Call } from "@siphon-project/control";

const MODE = (process.env.SIPHON_CONTROL_MODE ?? "outbound").toLowerCase();
const APP_NAME = process.env.SIPHON_CONTROL_APP ?? "ivr-app";
const TOKEN = process.env.IVR_APP_TOKEN ?? "changeme-dev-token";
const CONTROL_URL = process.env.SIPHON_CONTROL_URL ?? "ws://127.0.0.1:9092/control/ws";
const BIND = process.env.SIPHON_CONTROL_BIND ?? "127.0.0.1:8443";
const ANSWER_HOLD_MS = 5000;

/** The demo call flow: answer, stamp a per-call variable, hold, then hang up. */
async function handleCall(call: Call): Promise<void> {
  console.log(`[call] StasisStart ${call.channelId} sip_call_id=${call.sipCallId}`);
  try {
    await call.answer();
    // Per-call variables live on the control channel (drain with the call).
    await call.setVar("demo", "1");
    const demo = await call.getVar("demo");
    console.log(`[call] answered ${call.channelId}; demo=${demo}; holding ${ANSWER_HOLD_MS}ms`);
    await new Promise((resolve) => setTimeout(resolve, ANSWER_HOLD_MS));
    // Blind transfer instead of hanging up? -> await call.transfer("sip:agent@pbx")
    await call.hangup(); // alias for call.terminate()
    console.log(`[call] hung up ${call.channelId}`);
  } catch (error) {
    if (error instanceof ControlError) {
      // A dead/unknown call is a typed `not_found`; a verb the server does not
      // implement yet is `unsupported_verb`. Never fatal to the other calls.
      console.log(`[call] ${call.channelId} rejected: ${error.code}`);
    } else {
      throw error;
    }
  }
}

/** Inbound persistent: dial siphon, hello, resync, then drive assigned calls. */
async function runInbound(): Promise<void> {
  const client = await SipClient.connect({ url: CONTROL_URL, app: APP_NAME, token: TOKEN });
  console.log(`[control] connected (inbound) to ${CONTROL_URL} as ${APP_NAME}`);
  await client.onCall(handleCall); // drives reconnect + resync to completion
}

/** Outbound per-call-connect: siphon dials this server once per handed-over call. */
async function runOutbound(): Promise<void> {
  const [host, port] = BIND.split(":");
  const server = await SipServer.bind({
    host: host ?? "127.0.0.1",
    port: Number(port ?? 8443),
    app: APP_NAME,
    token: TOKEN,
  });
  console.log(`[control] listening (outbound per-call-connect) on ws://${BIND}`);
  await server.onCall(handleCall);
}

async function main(): Promise<void> {
  if (MODE === "inbound") {
    await runInbound();
  } else if (MODE === "outbound") {
    await runOutbound();
  } else {
    throw new Error(`SIPHON_CONTROL_MODE must be 'outbound' or 'inbound' (got ${MODE})`);
  }
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});

File: examples/remote_control/control_client.py
#!/usr/bin/env python3
"""Example external control application for the siphon control plane — Python SDK.

Drives live B2BUA calls a script hands over with ``call.handover("ivr-app")`` (the
ARI *Stasis* model) over siphon's control WebSocket, using the ``siphon-control``
Python SDK. The SDK owns the wire completely: no manual JSON, no ``rpc()`` helper,
no request-id bookkeeping — you get a ``Call`` whose verbs read like an in-process
siphon script (``call.answer()`` / ``call.transfer(...)`` / ``call.hangup()``).

Two connection modes, same wire (subprotocol ``siphon-control.v1``), same
``@on_call`` decorator, same ``Call`` handle — only the transport differs:

  - **outbound per-call-connect (the default)** — this app is a WebSocket
    *server* (``ControlServer``); siphon dials it once per handed-over call and
    the accepting socket owns that call. No ``hello`` — the first frame is
    ``StasisStart``. Use this for multi-pod / autoscaled controllers (siphon
    always dials *out*, so the "which pod owns the call" affinity problem cannot
    arise).
  - **inbound persistent** — this app is a WebSocket *client* (``ControlClient``)
    that connects in to ``control.listen`` and owns calls assigned to it. It
    sends a ``hello`` and ``resync``s to re-attach its calls after a reconnect.

Select the mode with ``SIPHON_CONTROL_MODE`` (``outbound`` | ``inbound``).

The demo answers each handed-over call, stamps a per-call variable, holds
briefly, then hangs up. Calls that are not handed over are unaffected.

Install the SDK::

    pip install siphon-control            # once published
    # ...or from this repo, into the active venv:
    #   cd siphon-control-sdk/siphon-control && maturin develop

Run::

    # outbound (default): siphon dials this server
    SIPHON_CONTROL_BIND=0.0.0.0:8443 IVR_APP_TOKEN=changeme-dev-token python control_client.py
    # inbound: this app dials siphon
    SIPHON_CONTROL_MODE=inbound IVR_APP_TOKEN=changeme-dev-token python control_client.py

See README.md for the matching siphon ``control:`` config and handover script.
"""
from __future__ import annotations

import asyncio
import os

from siphon_control import Call, ControlClient, ControlError, ControlServer

MODE = os.environ.get("SIPHON_CONTROL_MODE", "outbound").lower()
APP_NAME = os.environ.get("SIPHON_CONTROL_APP", "ivr-app")
TOKEN = os.environ.get("IVR_APP_TOKEN", "changeme-dev-token")
CONTROL_URL = os.environ.get("SIPHON_CONTROL_URL", "ws://127.0.0.1:9092/control/ws")
BIND = os.environ.get("SIPHON_CONTROL_BIND", "127.0.0.1:8443")

ANSWER_HOLD_SECONDS = 5.0


async def handle_call(call: Call) -> None:
    """Answer, stamp a per-call variable, hold, then hang up."""
    print(f"[call] StasisStart {call.channel_id} sip_call_id={call.sip_call_id}")
    try:
        await call.answer()
        # Per-call variables live on the control channel (drain with the call).
        await call.set_var("demo", "1")
        demo = await call.get_var("demo")
        print(f"[call] answered {call.channel_id}; demo={demo}; "
              f"holding {ANSWER_HOLD_SECONDS}s")
        await asyncio.sleep(ANSWER_HOLD_SECONDS)
        # Blind transfer instead of hanging up? -> await call.transfer("sip:agent@pbx")
        await call.hangup()
        print(f"[call] hung up {call.channel_id}")
    except ControlError as error:
        # A dead/unknown call is a typed `not_found`; a verb the server does not
        # implement yet is `unsupported_verb`. Never fatal to the other calls.
        print(f"[call] {call.channel_id} rejected: {error.code}")


async def run_inbound() -> None:
    """Inbound persistent: dial siphon, hello, resync, then drive assigned calls."""
    client = ControlClient(app=APP_NAME, token=TOKEN, url=CONTROL_URL)
    client.on_call(handle_call)
    print(f"[control] connecting (inbound) to {CONTROL_URL} as {APP_NAME!r}")
    # run() connects + hello, then drives the supervised reconnect + resync loop,
    # dispatching each handed-over call to handle_call.
    await client.run()


async def run_outbound() -> None:
    """Outbound per-call-connect: siphon dials this server once per handed-over call."""
    server = ControlServer(app=APP_NAME, token=TOKEN, bind=BIND)
    server.on_call(handle_call)
    # bind() resolves to the bound address (bind to `…:0` to learn an ephemeral
    # port before siphon dials in); serve() then accepts one call per dial forever.
    await server.bind()
    print(f"[control] listening (outbound per-call-connect) on ws://{server.local_addr}")
    await server.serve()


async def main() -> None:
    if MODE == "inbound":
        await run_inbound()
    elif MODE == "outbound":
        await run_outbound()
    else:
        raise SystemExit(
            f"SIPHON_CONTROL_MODE must be 'outbound' or 'inbound' (got {MODE!r})"
        )


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n[control] interrupted")
